napi_memory_management.md

```diff
@@ -0,0 +1,131 @@
+objects 'live' until they are no longer required by the native code,
+otherwise the objects could be collected before the native code was
+finished using them.
+
+As objects references are returned they are associated with a
+'scope'. The lifespan for the default scope is tied to the lifespan
+of the native method call. The result is that, by default, references
+remain valid and the objects associated with these references will be
+held live for the lifespan of the native method call.   
+
+In many cases, however, it is necessary that the references remain valid for
+either a shorter or longer lifespan than that of the native method.
+The sections which follow describe the N-API functions than can be used
+to change the reference lifespan from the default.
+
+## Making reference lifespan shorter than that of native method
+
+It is often necessary to make the lifespan for references shorter than
+the lifespan of a native method. For example, consider a native method
+that has a loop which iterates through the elements in a large array:
+
+```C
+  for (int i = 0; i < 1000000; i++) {
+    napi_value result;
+    napi_get_element(e object, i, &result);
+    // do something with element
+  }
+```
+
+This would result in a large number of references being created, consuming
+substantial resources. In addition, even through the native code could only
+use the most recent reference, all of the associated objects would also be
+kept alive, even though the native method was no longer able to access them.
+
+To handle this case, N-API provides the ability to establish a new 'scope' to
+which newly created references will be associated.  Once those references
+are no longer required, the scope can be 'closed' and any references associated
+with the scope are invalidated.  The methods available to open/close scopes are:
+
+```C
+NODE_EXTERN napi_status napi_open_handle_scope(napi_env e,
+                                               napi_handle_scope* result);
+NODE_EXTERN napi_status napi_close_handle_scope(napi_env e,
+                                                napi_handle_scope scope);
+```
+
+N-API only supports a single nested hiearchy of scopes.  There is only one
+active scope at any time, and all new references will be associated with that
+scope while it is active.  Scopes must be closed in the reverse order from
+which they are opened.  In addition, all scopes created within a native method
+must be closed before returning from that method.
+
+Taking the earlier example, adding calls to ```napi_open_handle_scope()``` and
+```napi_close_handle_scope()```  would ensure that at most a single reference
+is valid throughout the execution of the loop:
+
+```
+for (int i = 0; i < 1000000; i++) {napi_
+  napi_handle_scope scope;
+  napi_open_handle_scope(env, &scope);
+  napi_value result;
+  napi_get_element(e object, i, &result);
+  // do something with element
+  napi_close_handle_scope(env, scope);
+}
```

Comment:
The examples should take into account error handling. What if a N-API call returns non-`napi_ok`?